### DISCUSSION TITLE
Add info or opt-in warning for unused return values

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5748] [Language Ideas]
Basically, encourage the use of discard `_` to explicitly state that the return value of a method is not needed. I've just run across a bug caused by calling the wrong value-returning method, whose name was similar to another void-returning method's. Thoughts?

**Comment by HaloFour** [comment]
Sounds like good analyzer fodder.  IIRC R# does do this by default.

**Comment by Joe4evr** [comment]
~~[CA1806](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806) `Do not ignore method results`~~ [IDE0058](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0058) `Expression value is never used` already exists and ships in the SDK.

**Reply by TahirAhmadov** [reply]
Interesting! It has all these limitations - the return value has to be a string, or it's a LINQ method, or a p-invoke, etc. Why not apply this to all return values?

**Reply by Joe4evr** [reply]
I think the documentation is just a bit out of date. I've had quite a few cases of this diagnostic coming up lately and none of them were any of those.

**Reply by TahirAhmadov** [reply]
OK here's how it works for me:
C# 9, .NET FW 4.8, VS 2019 16.8.4 - this rule doesn't work at all. 
C# 9, .NET 5, VS 2019 16.11.3 - it's only applied as stated in the documentation (string, etc.). 
I'm wondering if it's a VS version difference? Or it depends on target framework?

**Reply by Joe4evr** [reply]
Ah, looks like I was a bit too fast on the trigger and it's `IDE0058`
![image](https://user-images.githubusercontent.com/1058562/153552142-b03c65ff-c58f-4eba-a53e-0084ddf526e8.png)

**Reply by TahirAhmadov** [reply]
Hmm, I installed the Microsoft.CodeAnalysis.NetAnalyzers package into the .NET 4.8 project, and added `<AnalysisMode>AllEnabledByDefault</AnalysisMode><AnalysisLevel>latest</AnalysisLevel>` to the project file (which is in SDK format), and that brought the CA1806, but it didn't bring IDE0058. Do I need to do anything special to enable it?
